### PR TITLE
v-rmanda/210208/classteams_with_assignments_powershellscript

### PR DESCRIPTION
**Pull request by v-rmanda**
Added a powershell script and ReadMe.md file

Instructions on how to run the script are given in the ReadMe.md file.

Powershell script generates a text file with list of classes with assignments of a tenant.

### CODE REVIEW

File: samples/powershell/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Powershell script
+
+This project has a script to list classes with assignments in a tenant
+
+## Prerequisites
+
+### Run the following commands in powershell as administrator
+* Install-Module Microsoft.Graph
+* Install-Module Microsoft.Graph -AllowClobber -Force
```
Comment (cristobal-buenrostro) [inline samples/powershell/README.md]:
```suggestion
* `Install-Module Microsoft.Graph`
* `Install-Module Microsoft.Graph -AllowClobber -Force`
```

File: samples/powershell/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+## Steps to run the script
+1. Open the script and give the below details of the client, username, password and save the file.
+-	    $tenantId = "YOUR_TENANT_ID"
+-	    $clientId = "YOUR_CLIENT_ID"
+-	    $username = user@example.com  # User's username (email or username)
+-	    $password = "UserPassword123"  # User's password (plaintext, should ideally be securely stored).
```
Comment (cristobal-buenrostro) [inline samples/powershell/README.md]:
```suggestion
-	    $username = "YOUR_USER_EMAIL"
-	    $password = "YOUR_USER_PASSWORD"
```

File: samples/powershell/README.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+2. Open PowerShell in administrator mode.	
+2. Navigate to location of drive where the above file is saved.
+3. Run the command – PowerShell -file <filename>.ps1
```
Comment (cristobal-buenrostro) [inline samples/powershell/README.md]:
```suggestion
3. Run the command `PowerShell -file <filename>.ps1`
```

File: samples/powershell/listclasseswithassignments.ps1.ps1
Comment (cristobal-buenrostro) [inline samples/powershell/listclasseswithassignments.ps1.ps1:1]: * use "-" to have a clear name: "list-classes-with-assignments.ps1"

* double check, you added **.ps1** twice in the file name

File: samples/powershell/listclasseswithassignments.ps1.ps1 [hunk 4]
```diff
@@ -0,0 +1,88 @@
+#prerequisite permissions
```
Comment (cristobal-buenrostro) [inline samples/powershell/listclasseswithassignments.ps1.ps1]:
```suggestion
# Prerequisite permissions
```

File: samples/powershell/listclasseswithassignments.ps1.ps1 [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+#Install-Module Microsoft.Graph
```
Comment (cristobal-buenrostro) [inline samples/powershell/listclasseswithassignments.ps1.ps1]:
```suggestion
# Install-Module Microsoft.Graph
```

File: samples/powershell/listclasseswithassignments.ps1.ps1 [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+#Install-Module Microsoft.Graph -AllowClobber -Force
```
Comment (cristobal-buenrostro) [inline samples/powershell/listclasseswithassignments.ps1.ps1]:
```suggestion
# Install-Module Microsoft.Graph -AllowClobber -Force
```

File: samples/powershell/listclasseswithassignments.ps1.ps1 [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+    # Make the request to get the access token
+    try {
+        $response = Invoke-RestMethod -Uri $tokenUrl -Method Post -ContentType "application/x-www-form-urlencoded" -Body $body
+        return $response.access_token
+    } catch {
+        Write-Error "Failed to obtain access token: $_"
+    }
+}
+# Get the access token
```
Comment (cristobal-buenrostro) [inline samples/powershell/listclasseswithassignments.ps1.ps1]:
```suggestion
}

# Get the access token
```

File: samples/powershell/listclasseswithassignments.ps1.ps1 [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+# Loop through each page of the results (handle pagination)
+do {
+	try {
+    # Get list of classes
+    $response = Invoke-RestMethod -Uri $uri -Headers @{Authorization = "Bearer $token"} -Method Get
+	} catch {
+        Write-Error "Failed to retrieve data from Graph API: $_"
+        break
+    }
```
Comment (cristobal-buenrostro) [inline samples/powershell/listclasseswithassignments.ps1.ps1]: nit: fix the tabulation for this block

File: samples/powershell/listclasseswithassignments.ps1.ps1 [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    # Loop through each class
+    foreach ($class in $response.value) {
+        $className = $class.displayName
+        $classId = $class.id
+
+        # Get assignments for the current class (with count=true)
+        $assignmentsUri = "https://graph.microsoft.com/beta/education/classes/$classId/assignments?$count=true"
+                try {
+            $assignmentsResponse = Invoke-RestMethod -Uri $assignmentsUri -Headers @{Authorization = "Bearer $token"} -Method Get -ErrorAction Stop
+        }
```
Comment (cristobal-buenrostro) [inline samples/powershell/listclasseswithassignments.ps1.ps1]: nit: fix the tabulation

File: samples/powershell/listclasseswithassignments.ps1.ps1 [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        catch {
+            Write-Host "Error retrieving assignments for class $className. Skipping this request."
+            continue  # Skip this class and move to the next class
+        }
+		
+		# Check if there are any assignments
```
Comment (cristobal-buenrostro) [inline samples/powershell/listclasseswithassignments.ps1.ps1]: nit: fix the tabulation for this comment

**Review by cristobal-buenrostro — CHANGES_REQUESTED** [review]

File: samples/powershell/listclasseswithassignments.ps1.ps1
Comment (v-rmanda) [reply samples/powershell/listclasseswithassignments.ps1.ps1:1]: Updated the file name to "list-classes-with-assignments.ps1"

Removed the additional ps1 from the file name

File: samples/powershell/listclasseswithassignments.ps1.ps1 (hunk 8 above)
Comment (v-rmanda) [reply samples/powershell/listclasseswithassignments.ps1.ps1]: fixed the tabulation for this block

File: samples/powershell/listclasseswithassignments.ps1.ps1 (hunk 10 above)
Comment (v-rmanda) [reply samples/powershell/listclasseswithassignments.ps1.ps1]: fixed the tabulation for try keyword to match with Catch block continue.

File: samples/powershell/listclasseswithassignments.ps1.ps1 (hunk 9 above)
Comment (v-rmanda) [reply samples/powershell/listclasseswithassignments.ps1.ps1]: fixed the tabulation

File: samples/powershell/list-classes-with-assignments.ps1 [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        $assignmentsUri = "https://graph.microsoft.com/beta/education/classes/$classId/assignments?$count=true"
+        try {
+            $assignmentsResponse = Invoke-RestMethod -Uri $assignmentsUri -Headers @{Authorization = "Bearer $token"} -Method Get -ErrorAction Stop
+        }
+        catch {
+            Write-Host "Error retrieving assignments for class $className. Skipping this request."
+            continue  # Skip this class and move to the next class
+        }
+		
+		# Check if there are any assignments
```
Comment (cristobal-buenrostro) [inline samples/powershell/list-classes-with-assignments.ps1]: nit: tabulation for this comment

**Review by cristobal-buenrostro — CHANGES_REQUESTED** [review]

File: samples/powershell/list-classes-with-assignments.ps1 (hunk 11 above)
Comment (v-rmanda) [reply samples/powershell/list-classes-with-assignments.ps1]: updated tab space with white space

**Review by cristobal-buenrostro — APPROVED** [review]

**Pull request merged**